ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,125 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
-            output_shape = arg0_shape_update;
+            // i.e. arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
+            NGRAPH_CHECK(Dimension::merge(
+                             merged_dimension, arg1_shape_tmp[0], arg0_shape_tmp[arg0_rank - 1]),
+                         "Incompatible matrix dimensions.");
+            arg0_shape_tmp.erase(arg0_shape_tmp.begin() + arg0_rank - 1);
+            return PartialShape(arg0_shape_tmp);
         }
-        else if (arg0_rank == 2 && arg1_rank == 2)
-        {
-            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
-            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
-        }
-        else
+
+        // 2D and bigger tensors cases.
+        NGRAPH_CHECK(Dimension::merge(merged_dimension,
+                                      arg0_shape_tmp[arg0_rank - 1],
+                                      arg1_shape_tmp[arg1_rank - 2]),
+                     "Incompatible matrix dimensions.");
+
+        auto max_rank = std::max(arg0_rank, arg1_rank);
+        std::vector<Dimension> output_shape(max_rank);
+
+        // Handle batch size
+        if (max_rank > 2)
         {
-            NGRAPH_CHECK(arg0_shape_update[arg0_rank - 1] == arg1_shape_update[arg1_rank - 2],
-                         "Incompatible arg shapes");
+            std::vector<Dimension> low_size_matrix =
+                arg0_rank > arg1_rank ? arg1_shape_tmp : arg0_shape_tmp;
+            std::vector<Dimension> big_size_matrix =
+                arg0_rank > arg1_rank ? arg0_shape_tmp : arg1_shape_tmp;
 
-            const auto& broadcast_shapes = builder::get_numpy_broadcast_shapes(
-                {Shape{begin(arg0_shape_update), next(end(arg0_shape_update), -2)},
-                 Shape{begin(arg1_shape_update), next(end(arg1_shape_update), -2)}});
+            if (arg0_rank != arg1_rank)
+            {
+                // Expand low_size_matrix (with 1) to have the same rank as big_size_matrix
+                size_t delta_rank = big_size_matrix.size() - low_size_matrix.size();
+                low_size_matrix.insert(low_size_matrix.begin(), delta_rank, 1);
+            }
 
-            output_shape = broadcast_shapes.first;
-            output_shape.insert(output_shape.end(), arg0_shape_update[arg0_rank - 2]);
-            output_shape.insert(output_shape.end(), arg1_shape_update[arg1_rank - 1]);
+            // Get max value for all batches (max_rank - 2), COL_INDEX_DIM and ROW_INDEX_DIM are
+            // updated at the end
+            for (auto i = 0; i < max_rank - 2; i++)
+            {
+                if (low_size_matrix[i].is_dynamic() || big_size_matrix[i].is_dynamic())
+                {
+                    // Static value is assigned to output when it is > 1, otherwise
+                    // dynamic dimension is forwarded to output
+                    Dimension::merge(merged_dimension, low_size_matrix[i], big_size_matrix[i]);
```

Comment:
Consider this valid case: both dimensions are dynamic, one has range [1..5], another one is [10..20]. Merge will return false as these ranges don't intersect. But 1 is included into the first range, so broadcasting of it is possible. Then it is a valid case and the output range should be [10..20].

Reply:
So you need to check if 1 is included into the range, and proceed it in a special way.

Reply:
@mitruska, you convinced me that handling ranges for Rank is not possible with the current nGraph code, but for each dimension separately it is still possible and we have to make it as precise as possible.

Reply:
Ok, I thought that we want to stay with fully dynamic dimensions. 
I've just added handling dynamic dimensions range. Please take a look.